validation: assumeutxo params mainnet

Sjors opened the pull request:
This adds snapshot parameters for mainnet block 840,000.

You can generate the snapshot yourself using `./contrib/devtools/utxo_snapshot.sh` or download my torrent:
* torrent: `magnet:?xt=urn:btih:596c26cc709e213fdfec997183ff67067241440c&dn=utxo-840000.dat&tr=udp%3A%2F%2Ftracker.bitcoin.sprovoost.nl%3A6969`

It would be a good idea to test:
1. That you can produce the same snapshot file, sha256 sum:

```
dc4bb43d58d6a25e91eae93eb052d72e3318bd98ec62a5d0c11817cefbba177b utxo-840000.dat
```

2. That the snapshot works

Sjors commented:
Updated `hash_serialized` after #28685.

I only tested that the `loadtxoutset` command finishes.

@pablomartin4btc I found a different hash than you... https://github.com/Sjors/bitcoin/commit/24deb2022b822f22fba9fcbee201e37a83225eb2#commitcomment-130847775

pablomartin4btc reviewed (approved):
tACK 0800be24bc6de4154653be75b8fa9902a4a8c3a7

```
./contrib/devtools/utxo_snapshot.sh 800000 - ./src/bitcoin-cli -datadir=${AU_DATADIR}
Rewinding chain back to height 800000 (by invalidating 00000000000000000000e26b239cf19ec7ace5edd9694d51a3f6933247720947); this may take a while
Generating txoutset info...
{
  "height": 800000,
  "bestblock": "00000000000000000002a7c4c1e48d76c5a37902165a270156b7a8d72728a054",
  "txouts": 111535121,
  "bogosize": 8443325590,
  "hash_serialized_3": "6e63799cac0ab45f94789f97692caf691ee66625e920cbf2525fceb24509adda",
  "total_amount": 19437286.83643360,
  "transactions": 72936026,
  "disk_size": 7264997650
}
Restoring chain to original height; this may take a while
```

I incorrectly took the `muhash` value from `gettxoutsetinfo` (without running the `utxo_snapshot.sh` which also performs `invalidateblock` right before).

Sjors commented:
@pablomartin4btc thanks for checking that you get the same snapshot. In addition it would be good to test if you're able to load the snapshot and complete the background sync.

pablomartin4btc commented:
> @pablomartin4btc thanks for checking that you get the same snapshot. In addition it would be good to test if you're able to load the snapshot and complete the background sync.

I've already checked IBD completed fine with`[background validation]`finished, but with the previous`hash_serialised`, as I wanted to verify that #28698 was only happening before the IBD completion.

I'll run all the process again with the new`hash_serialised`value.

pablomartin4btc reviewed (commented):
Tested `loadtxoutset` a few times (pending IBD to finish) with `coinstatsindex` & `blockfilterindex`,  and without them. I've also validated that the `sha256sum` corresponds to the one obtained from the downloaded file. 

Now, having a pruned node with IBD completed, from which I got the `m_assumeutxo_data` mentioned in my previous review, when I run `./contrib/devtools/utxo_snapshot.sh` with a filename in order to perform the actual `dumptxoutset`within the script, I got a node crash with `ERROR: ReadBlockFromDisk: OpenBlockFile failed for FlatFilePos(nFile=-1, nPos=0)`, which instance I had to kill since it was unresponsive. I'm not sure if this situation was due to the `invalidateblock` performed twice by runnning the `utxo_snapshot` script (one to get the info from `gettxoutsetinfo` and the second for the actual dump), the pruned node or a combination of all three. Starting the node again without `blockfilterindex` (option given by the node before doing the suggested `-reindex` or `-reindex-chainstate` which doesn't work on pruned nodes), made it worse till getting `!setBlockIndexCandidates.empty()`. I see there are a few open issues regarding `invalidateblock` & `setBlockIndexCandidates` so before investigating further I'd like to know first if the `dumptxoutset` is expected to work from a fully synced pruned node?

Sjors commented:
As discussed offline, the `utxo_snapshot.sh` script should have a sanity check to make the node isn't pruned below the snapshot height. On a pruned node you can only make a snapshot for a block that hasn't been pruned yet.

pablomartin4btc commented:
> As discussed offline, the `utxo_snapshot.sh` script should have a sanity check to make the node isn't pruned below the snapshot height. On a pruned node you can only make a snapshot for a block that hasn't been pruned yet.

I agree with it, mentioned on a #27845 that's touching the script ATM.

In the testing I performed and mentioned above the pruned node was above the snapshot height (`800'000`):

```
  "pruned": true,
  "pruneheight": 600510,
  "automatic_pruning": true,
  "prune_target_size": 20971520000,

```

Sjors commented:
I plan to make a fresh snapshot after #26045 lands.

fjahr commented:
Didn't see this one until the conflict popped up. I think we can move to 830,000 for v27.0 and I am unsure if we need to wait for #26045. I will bring it up in the next meeting.

Sjors commented:
I switched to the halving block! Also dropped all prerequisites but #26045 from the description, but see conceptual discussion in #29616.

Sjors commented:
Rebased now that #29612 landed. Will do some final testing before marking this ready to review.

fjahr commented:
tACK f193b0fd853cb9e2f7cb4ed9dd6fb5d2662a04fe

I have re-tested the full assumeutxo flow using this snapshot: Downloaded the snapshot from the torrent link, verified the hash, successfully loaded the snapshot in a fresh node, verified that the snapshot chainstate was able to sync to the tip and the node was usable. Then waited until the background chainstate caught up to the snapshot and that was also successful. Restarted the node and verified the cleanup of the second chainstate. Also checked that I get the exact same dump using `dumptxoutset` with #29553 on a fully synced node that did not use the snapshot to sync.

achow101 commented:
Looks like there's a silent merge conflict:

```
kernel/chainparams.cpp: In constructor ‘CMainParams::CMainParams()’:
kernel/chainparams.cpp:194:9: error: no match for ‘operator=’ (operand types are ‘std::__debug::vector<AssumeutxoData>’ and ‘<brace-enclosed initializer list>’)
  194 |         };
      |         ^
In file included from /usr/include/c++/11/vector:76,
                 from /usr/include/c++/11/functional:62,
                 from /usr/include/c++/11/pstl/glue_algorithm_defs.h:13,
                 from /usr/include/c++/11/algorithm:74,
                 from ./span.h:8,
                 from ./uint256.h:10,
                 from ./consensus/params.h:9,
                 from ./kernel/chainparams.h:9,
                 from kernel/chainparams.cpp:6:
/usr/include/c++/11/debug/vector:249:7: note: candidate: ‘std::__debug::vector<_Tp, _Allocator>& std::__debug::vector<_Tp, _Allocator>::operator=(const std::__debug::vector<_Tp, _Allocator>&) [with _Tp = AssumeutxoData; _Allocator = std::allocator<AssumeutxoData>]’
  249 |       operator=(const vector&) = default;
      |       ^~~~~~~~
/usr/include/c++/11/debug/vector:249:17: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘const std::__debug::vector<AssumeutxoData>&’
  249 |       operator=(const vector&) = default;
      |                 ^~~~~~~~~~~~~
/usr/include/c++/11/debug/vector:252:7: note: candidate: ‘std::__debug::vector<_Tp, _Allocator>& std::__debug::vector<_Tp, _Allocator>::operator=(std::__debug::vector<_Tp, _Allocator>&&) [with _Tp = AssumeutxoData; _Allocator = std::allocator<AssumeutxoData>]’
  252 |       operator=(vector&&) = default;
      |       ^~~~~~~~
/usr/include/c++/11/debug/vector:252:17: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘std::__debug::vector<AssumeutxoData>&&’
  252 |       operator=(vector&&) = default;
      |                 ^~~~~~~~
/usr/include/c++/11/debug/vector:255:7: note: candidate: ‘std::__debug::vector<_Tp, _Allocator>& std::__debug::vector<_Tp, _Allocator>::operator=(std::initializer_list<_Tp>) [with _Tp = AssumeutxoData; _Allocator = std::allocator<AssumeutxoData>]’
  255 |       operator=(initializer_list<value_type> __l)
      |       ^~~~~~~~
/usr/include/c++/11/debug/vector:255:46: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘std::initializer_list<AssumeutxoData>’
  255 |       operator=(initializer_list<value_type> __l)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
```

Sjors committed:
chainparams: add mainnet assumeutxo param at height 840_000

fjahr commented:
re-ACK fb543c9f733610b8fadbfd6f1c71f1bb520de985

Only the merge conflict was addressed, the actual values did not change, per `git range-diff master f193b0fd853cb9e2f7cb4ed9dd6fb5d2662a04fe fb543c9f733610b8fadbfd6f1c71f1bb520de985`. Also confirmed that the dump can still be loaded without error.

Sjors commented:
Rebased after the (tiny) format change in #30598, for easier testing. Updated torrent link and snapshot file hash in the description. The chainparams themselves are unchanged.

fjahr commented:
re-ACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410

Verified only rebase via `git range-diff master fb543c9f733610b8fadbfd6f1c71f1bb520de985 1610643c8b37a9f674b236cfa79abf8f8aaf1410` and re-tested loading and syncing with the new snapshot torrent.

Sjors commented:
```
2024-08-10T19:39:49Z [snapshot] computing UTXO stats for background chainstate to validate snapshot - this could take a few minutes
Killed
```

I think this happened because I used a RAM drive for pruned sync and it ran out of space (a Guix build I ran at the same time failed too). Would be nice to fail more gracefully, but I don't think it's related to the specific snapshot here.

achow101 commented:
ACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410

If a few more people can review this in the next day or so, I believe we can have this in for 28.0.

fjahr commented:
> If a few more people can review this in the next day or so, I believe we can have this in for 28.0.

I have pinged a few more people for (re-)reviews and testing but please give it until the end of the week because testing the full flow can take a while depending on what machine you are doing it on or what connection speed you have.

theStack reviewed (approved):
Tested ACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410

Loaded the provided snapshot on a freshly created pruned node (`-prune=20000`), and waited first until the snapshot chainstate left IBD, then until the background validation was finished successfully. Right after loading, I also checked that UTXO-querying RPCs like `gettxoutsetinfo` / `scantxoutset` / `gettxout` use the chainstate snapshot (see also PR #30636). Also created a dump on another node on height 840000 and verified that the AssumeUtxo hash (`"txoutsethash"` result of `dumptxoutset`) and the snapshot file sha256 hashes match. ✔️ Didn't test any wallet functionality though.

Btw, if someone needs a simple torrent downloading tool for the CLI, I can recommend https://aria2.github.io/ (just don't forget to put the magnet URI between single quotes, i.e. `$ aria2c 'magnet:...'`; without them the link in the PR description is still accepted, but the download won't start).

mzumsande reviewed (commented):
tested ACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410

I re-created the snapshot at height 840k and verified that the file is identical to the one I downloaded from the torrent.
I also successfully loaded the snapshot into a new node (though I didn't complete the sync).

willcl-ark reviewed (approved):
tACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410

Checked that the snapshot was created and had the same `sha256sum`

<details>
<summary>Details</summary>

```log
$ ./contrib/devtools/utxo_snapshot.sh 840000 snapshot.dat ./src/bitcoin-cli
Do you want to disable network activity (setnetworkactive false) before running invalidateblock? (Y/n): y
Disabling network activity
false
Rewinding chain back to height 840000 (by invalidating 00000000000000000001b48a75d5a3077913f3f441eb7e08c13c43f768db2463); this may take a while
Generating UTXO snapshot...
{
  "coins_written": 176948713,
  "base_hash": "0000000000000000000320283a032748cef8227873ff4872689bf23f1cda83a5",
  "base_height": 840000,
  "path": "/home/will/.bitcoin/snapshot.dat",
  "txoutset_hash": "a2a5521b1b5ab65f67818e5e8eccabb7171a517f9e2382208f77687310768f96",
  "nchaintx": 991032194
}
Restoring chain to original height; this may take a while
# ....

$ sha256sum snapshot.dat
dc4bb43d58d6a25e91eae93eb052d72e3318bd98ec62a5d0c11817cefbba177b  snapshot.dat
```

</details>

I also loaded the snapshot into a fresh node which worked (but didn't yet complete the sync).

pablomartin4btc reviewed (commented):
tACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410

Checksum matches downloaded file and performed `loadtxoutset` on a new node.

I'm still performing some other tests.

pablomartin4btc commented:
<details>
<summary>Continued testing on a full node until validation is finished (<code>getchainstate</code>) and fully synced, also on a pruned node (<code>-prune=30000</code>).</summary>

```
./src/bitcoin-cli -datadir=${AU_DATADIR} getchainstates
{
  "headers": 857301,
  "chainstates": [
    {
      "blocks": 857301,
      "bestblockhash": "000000000000000000017a865975aac40cdb7d05f4010a6e915ffe7af260a151",
      "difficulty": 86871474313761.95,
      "verificationprogress": 0.9999994926046412,
      "coins_db_cache_bytes": 8388608,
      "coins_tip_cache_bytes": 17169383424,
      "snapshot_blockhash": "0000000000000000000320283a032748cef8227873ff4872689bf23f1cda83a5",
      "validated": true
    }
  ]
}

```
</details>

Replicated behaviour as [described](https://github.com/bitcoin/bitcoin/pull/28553#issuecomment-2282740748) by @Sjors and after freeing the necessary space the node continued running/ syncing successfully. 

Also tested the UTXO-querying RPCs as @theStack [above](https://github.com/bitcoin/bitcoin/pull/28553#pullrequestreview-2235126233).

<details>
<summary>I've generated the snapshot myself which matches the one provided by @Sjors at the top.</summary>

```
./contrib/devtools/utxo_snapshot.sh 840000 snapshot.dat ./src/bitcoin-cli -datadir=${AU_DATADIR}
Do you want to disable network activity (setnetworkactive false) before running invalidateblock? (Y/n): 
Disabling network activity
false
Rewinding chain back to height 840000 (by invalidating 00000000000000000001b48a75d5a3077913f3f441eb7e08c13c43f768db2463); this may take a while
Generating UTXO snapshot...
{
  "coins_written": 176948713,
  "base_hash": "0000000000000000000320283a032748cef8227873ff4872689bf23f1cda83a5",
  "base_height": 840000,
  "path": "/home/pablo/.test_utxo_840/snapshot.dat",
  "txoutset_hash": "a2a5521b1b5ab65f67818e5e8eccabb7171a517f9e2382208f77687310768f96",
  "nchaintx": 991032194
}
Restoring chain to original height; this may take a while
Restoring network activity
true


```
</details>

<details>

<summary><code>loadtxoutset</code> ouput I missed in my comment <a href="https://github.com/bitcoin/bitcoin/pull/28553#pullrequestreview-2238694990">above</a>.</summary>

```
./src/bitcoin-cli -datadir=${AU_DATADIR} loadtxoutset /home/pablo/.test_utxo_2/utxo-840000.dat 
{
  "coins_loaded": 176948713,
  "tip_hash": "0000000000000000000320283a032748cef8227873ff4872689bf23f1cda83a5",
  "base_height": 840000,
  "path": "/home/pablo/.test_utxo_2/utxo-840000.dat"
}

```
</details>

<details>
<summary><code>./contrib/devtools/utxo_snapshot.sh</code> has a bug which is that if we run the script without the node reaching the block yet it would disable the network if user selects that option.</summary>

```
/contrib/devtools/utxo_snapshot.sh 840000 snapshot2.dat ./src/bitcoin-cli -datadir=${AU_DATADIR}
Do you want to disable network activity (setnetworkactive false) before running invalidateblock? (Y/n): 
Disabling network activity
false
error code: -8
error message:
Block height out of range

```

And the user will see the following in the node and it would stay there if not reset:

```
2024-08-21T14:44:13Z UpdateTip: new best=00000000000000afa0cd000a16e244f56032735d41acd32ac00337aceb2a5240 height=235382 version=0x00000002 log2_work=69.987697 tx=17492185 date='2013-05-09T23:54:32Z' progress=0.016219 cache=71.0MiB(571085txo)
2024-08-21T14:44:13Z UpdateTip: new best=0000000000000087c5e0b820afff496b95ba44ad64640c73b234d3261d3f99d2 height=235383 version=0x00000002 log2_work=69.987750 tx=17492341 date='2013-05-09T23:54:47Z' progress=0.016219 cache=71.0MiB(571291txo)
2024-08-21T14:44:13Z UpdateTip: new best=000000000000014a4b5fddf3c8abb6209247255ca9e8df786b271dd1b2ac82a6 height=235384 version=0x00000002 log2_work=69.987804 tx=17492344 date='2013-05-10T00:20:18Z' progress=0.016219 cache=71.0MiB(571297txo)
2024-08-21T14:44:13Z SetNetworkActive: false

```

</details>

<details>
<summary>I've a fix already and will create a PR in a bit.</summary>

```
/contrib/devtools/utxo_snapshot.sh 840000 snapshot2.dat ./src/bitcoin-cli -datadir=${AU_DATADIR}
Error: The node has not yet synchronized to block height 840001.
Please wait until the node has synchronized past this block height and try again.
```

</details>

pablomartin4btc reviewed (commented):
<details>
<summary>Continued testing on a full node until validation is finished (<code>getchainstate</code>) and fully synced, also on a pruned node (<code>-prune=30000</code>).</summary>

```
./src/bitcoin-cli -datadir=${AU_DATADIR} getchainstates
{
  "headers": 857301,
  "chainstates": [
    {
      "blocks": 857301,
      "bestblockhash": "000000000000000000017a865975aac40cdb7d05f4010a6e915ffe7af260a151",
      "difficulty": 86871474313761.95,
      "verificationprogress": 0.9999994926046412,
      "coins_db_cache_bytes": 8388608,
      "coins_tip_cache_bytes": 17169383424,
      "snapshot_blockhash": "0000000000000000000320283a032748cef8227873ff4872689bf23f1cda83a5",
      "validated": true
    }
  ]
}

```
</details>

Replicated behaviour as [described](https://github.com/bitcoin/bitcoin/pull/28553#issuecomment-2282740748) by @Sjors and after freeing the necessary space the node continued running/ syncing successfully. 

Also tested the UTXO-querying RPCs as @theStack [above](https://github.com/bitcoin/bitcoin/pull/28553#pullrequestreview-2235126233).

<details>
<summary>I've generated the snapshot myself which matches the one provided by @Sjors at the top.</summary>

```
./contrib/devtools/utxo_snapshot.sh 840000 snapshot.dat ./src/bitcoin-cli -datadir=${AU_DATADIR}
Do you want to disable network activity (setnetworkactive false) before running invalidateblock? (Y/n): 
Disabling network activity
false
Rewinding chain back to height 840000 (by invalidating 00000000000000000001b48a75d5a3077913f3f441eb7e08c13c43f768db2463); this may take a while
Generating UTXO snapshot...
{
  "coins_written": 176948713,
  "base_hash": "0000000000000000000320283a032748cef8227873ff4872689bf23f1cda83a5",
  "base_height": 840000,
  "path": "/home/pablo/.test_utxo_840/snapshot.dat",
  "txoutset_hash": "a2a5521b1b5ab65f67818e5e8eccabb7171a517f9e2382208f77687310768f96",
  "nchaintx": 991032194
}
Restoring chain to original height; this may take a while
Restoring network activity
true


```
</details>

<details>

<summary><code>loadtxoutset</code> ouput I missed in my comment <a href="https://github.com/bitcoin/bitcoin/pull/28553#pullrequestreview-2238694990">above</a>.</summary>

```
./src/bitcoin-cli -datadir=${AU_DATADIR} loadtxoutset /home/pablo/.test_utxo_2/utxo-840000.dat 
{
  "coins_loaded": 176948713,
  "tip_hash": "0000000000000000000320283a032748cef8227873ff4872689bf23f1cda83a5",
  "base_height": 840000,
  "path": "/home/pablo/.test_utxo_2/utxo-840000.dat"
}

```
</details>

<details>
<summary><code>./contrib/devtools/utxo_snapshot.sh</code> has a bug which is that if we run the script without the node reaching the block yet it would disable the network if user selects that option.</summary>

```
/contrib/devtools/utxo_snapshot.sh 840000 snapshot2.dat ./src/bitcoin-cli -datadir=${AU_DATADIR}
Do you want to disable network activity (setnetworkactive false) before running invalidateblock? (Y/n): 
Disabling network activity
false
error code: -8
error message:
Block height out of range

```

And the user will see the following in the node and it would stay there if not reset:

```
2024-08-21T14:44:13Z UpdateTip: new best=00000000000000afa0cd000a16e244f56032735d41acd32ac00337aceb2a5240 height=235382 version=0x00000002 log2_work=69.987697 tx=17492185 date='2013-05-09T23:54:32Z' progress=0.016219 cache=71.0MiB(571085txo)
2024-08-21T14:44:13Z UpdateTip: new best=0000000000000087c5e0b820afff496b95ba44ad64640c73b234d3261d3f99d2 height=235383 version=0x00000002 log2_work=69.987750 tx=17492341 date='2013-05-09T23:54:47Z' progress=0.016219 cache=71.0MiB(571291txo)
2024-08-21T14:44:13Z UpdateTip: new best=000000000000014a4b5fddf3c8abb6209247255ca9e8df786b271dd1b2ac82a6 height=235384 version=0x00000002 log2_work=69.987804 tx=17492344 date='2013-05-10T00:20:18Z' progress=0.016219 cache=71.0MiB(571297txo)
2024-08-21T14:44:13Z SetNetworkActive: false

```

</details>

<details>
<summary>I've a fix already and will create a PR in a bit.</summary>

```
/contrib/devtools/utxo_snapshot.sh 840000 snapshot2.dat ./src/bitcoin-cli -datadir=${AU_DATADIR}
Error: The node has not yet synchronized to block height 840001.
Please wait until the node has synchronized past this block height and try again.
```

</details>
